implementation/src/main/java/io/smallrye/faulttolerance/SimpleCommand.java

```diff
@@ -85,15 +96,38 @@ protected Object run() throws Exception {
 
     @Override
     protected Object getFallback() {
+        if (fallback == null) {
+            return super.getFallback();
+        }
         Throwable failure = getFailedExecutionException();
         if (failure != null && operation.hasCircuitBreaker() && !isFailureAssignableFromAnyFailureException(failure)) {
             // Command failed but the fallback should not be used
             throw new FailureNotHandledException(failure);
         }
-        if (fallback == null) {
+        if (failure == null) {
+            failure = translateException();
+        }
+        if (retryContext == null || !retryContext.shouldRetryOn(failure)) {
+            return fallback.get();
+        } else {
             return super.getFallback();
         }
-        return fallback.get();
+    }
+
+    private Throwable translateException() {
+        Exception e = executionResult.getExecutionException();
```

Comment:
this isn't very good.

Reply:
Then we should probably leave a comment and create an issue...

Reply:
done